src/cqrs/aggregate-root.ts

```diff
@@ -55,10 +55,15 @@ export abstract class AggregateRoot<EventBase extends IEvent = IEvent> {
         this.publishers.length
       } publishers`,
     );
+
+    // flush the queue first to avoid multiple commit of the same event on concurrent calls
```

Comment:
what happens when one of the `await publisher(events);` fails? don't we loose events?

Reply:
Could be. But this is also an issue when calling multiple times the `commit()` method while the event queued to be published are kept unchanged:

```
 IDEMPOTENT WRITE TO STREAM ClientCorrelationID b9f8fb7f-0797-4abb-863c-1e54c8d38d3d, EventStreamId: googleshopping_sync-UUpKevdNkVW2CdkHauE3SvJBrUG3, CorrelationId: d6d0539e-58a4-4a91-aa8a-94515ea10a78, FirstEventNumber: 697, LastEventNumber: 700.
[00001,16,09:01:52.556] IDEMPOTENT WRITE TO STREAM ClientCorrelationID aac533d5-cc9d-4787-939b-f98f678c3a20, EventStreamId: googleshopping_sync-UUpKevdNkVW2CdkHauE3SvJBrUG3, CorrelationId: e20391d5-6e88-431e-b55c-88fe3d0781d3, FirstEventNumber: 697, LastEventNumber: 702.
[00001,16,09:01:52.691] IDEMPOTENT WRITE TO STREAM ClientCorrelationID a493ea57-84f4-4844-aae8-fa9e9a215fef, EventStreamId: googleshopping_sync-UUpKevdNkVW2CdkHauE3SvJBrUG3, CorrelationId: 670135c7-7bfa-4353-816c-9384ee497b44, FirstEventNumber: 697, LastEventNumber: 702.
[00001,14,09:01:53.139] Dropping idempotent write to stream "googleshopping_sync-UUpKevdNkVW2CdkHauE3SvJBrUG3", startEventNumber: 705, endEventNumber: 708 since the original write has not yet been replicated.
[00001,14,09:01:54.288] Dropping idempotent write to stream "googleshopping_sync-UUpKevdNkVW2CdkHauE3SvJBrUG3", startEventNumber: 709, endEventNumber: 710 since the original write has not yet been replicated.
```

And in this case you miss events (connexion dropped, next events lost).

Reply:
so you mean a situation where some events are stacked, ready to be commited, and then commit is called several times, and what you say is that at this moment we loose events?

Reply:
If you don't call commit asynchronously yes. Which is usually the case when you use `this.apply()` in your aggregate with autoCommit on.

Reply:
In the code we have : 
```typescript
  async apply<T extends EventBase = EventBase>(
    event: T,
    isFromHistory = false,
  ) {
    this.logger.debug(
      `Applying ${event.constructor.name} with${
        !!this.autoCommit ? '' : 'out'
      } autocommit`,
    );
    if (!isFromHistory) {
      this.addEvent(event);
    }
    this.autoCommit && (await this.commit());
    const handler = this.getEventHandler(event);
    handler && (await handler.call(this, event));
  }
```

This method is async, so why not simply call ` await this.apply() ` ?

Reply:
This way, no event loose, and no need to change the code, yes?

Reply:
https://github.com/PrestaShopCorp/nestjs-geteventstore/blob/4.0.1/examples/src/heroes/aggregates/hero.aggregate.ts#L15 no async await here in the `this.apply()` calls

Reply:
yes, probably because it's a basic example, it doesn't forbid to use an await

Reply:
of course using an async could solve this, but in 100% of our code using this lib this.apply is not called async.

Reply:
of course using an async could solve this, but in 100% of our code using this lib this.apply is not called async.

Reply:
So to sum up, here is are the options : 
1 - Take a risk to loose some events by doing several commit asynchronously
2 - Not taking this risk, but having to replace `this.apply` with `await this.apply`

What do you think about that ?

Reply:
I don't know, I'd like to see the implementation of option 3 to see you better

Reply:
https://github.com/PrestaShopCorp/nestjs-geteventstore/pull/88/commits/f9a655a956040284cf3643abccc6173ab24c7267

Reply:
@AlexandreDecollas would this fit your requirements now?

Reply:
`this.clearEvents();` is equivalent to `this[INTERNAL_EVENTS].length = 0` so doing `this[INTERNAL_EVENTS].unshift(...events);` after a `this.clearEvents();` is useless.

And if I'm not wrong, it doesn't change the fact that if one if the publisher fails, we loose event, yes?

Reply:
If the publisher fails we still requeue the events, so that will be fine: next commit will handle this.
Unshift is useful if some events are added in the meantime! Don't forget the key point here: commit can be called asynchronously and concurrently. The function can be 50% executed before another call is being triggered... and other events collected in the meantime.

Reply:
ok for me